### PR TITLE
[SPARK-48584][SQL] Perf improvement for unescapePathName

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR improves perf for unescapePathName with algorithms briefly described as:
- If a path contains no '%' or contains '%' at `position > path.length-2`, we return the original identity instead of creating a new StringBuilder to append char by char
- Otherwise, we loop with 2 indices, `plaintextStartIdx` which starts from 0 and then points to the next char after resolving `%xx`, and `plaintextEndIdx` which points to the next `'%'`. `plaintextStartIdx` moves to `plaintextEndIdx + 3` if `%xx` is valid, or moves to `plaintextEndIdx + 1` if `%xx` is invalid. 
- Instead of using Integer.parseInt with error capture, we identify the high and low characters manually. 

### Why are the changes needed?

performance improvement for hotspots

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

- new tests in ExternalCatalogUtilsSuite
- Benchmark results (9-11x faster)
### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @dongjoon-hyun @cloud-fan @LuciferYang @JoshRosen thanks

**Comment by beliefer** [comment]
Could you provide some micro benchmark?

**Comment by beliefer** [comment]
Oh, I see.

**Comment by LuciferYang** [comment]
It seems that there is a typo in the pr description:

```
Instead of using Integer.parseInt with error capture, we identify the high and low chararaters manually.
```

chararaters -> characters

**Comment by yaooqinn** [comment]
Thank you @LuciferYang

**Review by LuciferYang — APPROVED** [review]
+1, LGTM
Thanks @yaooqinn

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala [hunk 1]
```diff
@@ -97,31 +98,40 @@ object ExternalCatalogUtils {
...
-        }
-        if (code >= 0) {
-          sb.append(code.asInstanceOf[Char])
-          i += 3
+    if (path == null || path.isEmpty) {
+      return path
+    }
+    var plaintextEndIdx = path.indexOf('%')
+    val length = path.length
+    if (plaintextEndIdx == -1 || plaintextEndIdx + 2 > path.length) {
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala:106]: How about `if (plaintextEndIdx == -1 || plaintextEndIdx + 2 > path.length || path.lastIndexOf('%')) ` ?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala:106]: I'm sorry, I'm having a little trouble understanding your suggestion. Can you please clarify?

**Comment by yaooqinn** [comment]
Thank you @LuciferYang, I added a nit in the last commit and passed the test locally, so I didn't wait for the CI.


Merged to master

**Pull request closed without merging**